pkg/watch/versioned/register.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package versioned
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// WatchEventKind is name reserved for serializing watch events.
+const WatchEventKind = "WatchEvent"
+
+// AddToGroupVersion registers the watch external and internal kinds with the scheme, and ensures the proper
+// conversions are in place.
+func AddToGroupVersion(scheme *runtime.Scheme, groupVersion unversioned.GroupVersion) {
+	scheme.AddKnownTypeWithName(groupVersion.WithKind(WatchEventKind), &Event{})
+	scheme.AddKnownTypeWithName(
+		unversioned.GroupVersion{Group: groupVersion.Group, Version: runtime.APIVersionInternal}.WithKind(WatchEventKind),
+		&InternalEvent{},
+	)
+	scheme.AddConversionFuncs(
+		Convert_versioned_Event_to_watch_Event,
+		Convert_versioned_InternalEvent_to_versioned_Event,
+		Convert_watch_Event_to_versioned_Event,
+		Convert_versioned_Event_to_versioned_InternalEvent,
+	)
+}
+
+func Convert_watch_Event_to_versioned_Event(in *watch.Event, out *Event, s conversion.Scope) error {
+	out.Type = string(in.Type)
+	switch t := in.Object.(type) {
+	case *runtime.Unknown:
+		// TODO: handle other fields on Unknown and detect type
+		out.Object.Raw = t.Raw
+	case nil:
+	default:
+		out.Object.Object = in.Object
```

Comment:
In this case we're calling convert which does not guarantee the incoming
object is mutated.  Since runtime.Object is an interface and we use some
other types in interfaces that aren't purely serializable
(runtime.Encodable) it's not safe to copy.  I can make a comment in the
godoc of runtime.Extension that it does not perform a copy.

On Mon, Apr 11, 2016 at 4:15 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/watch/versioned/register.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r59167904>
> :
>
> > +		Convert_versioned_Event_to_watch_Event,
> > +		Convert_versioned_InternalEvent_to_versioned_Event,
> > +		Convert_watch_Event_to_versioned_Event,
> > +		Convert_versioned_Event_to_versioned_InternalEvent,
> > +	)
> > +}
> > +
> > +func Convert_watch_Event_to_versioned_Event(in *watch.Event, out *Event, s conversion.Scope) error {
> > +	out.Type = string(in.Type)
> > +	switch t := in.Object.(type) {
> > +	case *runtime.Unknown:
> > +		// TODO: handle other fields on Unknown and detect type
> > +		out.Object.Raw = t.Raw
> > +	case nil:
> > +	default:
> > +		out.Object.Object = in.Object
>
> Shouldn't this be a DeepCopy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/e6879a33f46577ce93328230a28453244c3a408f#r59167904>
>